docs/proposals/cross-mesh-ux.md

```diff
@@ -0,0 +1,229 @@
+# Cross-mesh communication UX
+
+_Issue_: https://github.com/kumahq/kuma/issues/3323
+
+This proposal allows the operators of one `Mesh` to expose services to other `Meshes`
+via `MeshGateway`, the same interface for exposing to external traffic.
+
+The proposal addresses UX both for _exposing_ and the _consuming_
+meshes and only HTTP/S services.
+
+## Design
+
+In the following examples, the mesh exposing its services will be `expose` and
+the mesh consuming another mesh's services will be `consume`.
+
+<details>
+<summary>Recap: <code>MeshGateway</code> configuration</summary>
+
+```
+type: MeshGateway
+mesh: expose
+name: edge-gateway
+selectors:
+- match:
+    kuma.io/service: edge-gateway
+conf:
+  listeners:
+  - port: 8080
+    protocol: HTTP
+    hostname: foo.example.com
+    tags:
+      port: "8080"
+---
+type: MeshGatewayRoute
+mesh: expose
+name: edge-gateway-route
+selectors:
+  - match:
+      kuma.io/service: edge-gateway
+conf:
+  http:
+    rules:
+      - matches:
+          - path:
+              match: PREFIX
+              value: /api
+        backends:
+          - destination:
+              kuma.io/service: server
+```
+
+</details>
+
+### `expose` UX
+
+The main UX feature will be a mechanism to configure a
+`MeshGateway` to permit traffic from services in one
+`Mesh` to reach the services behind the `MeshGateway`.
+The configuration will make it possible to expose `MeshGatewayRoute`s depending
+on cross-mesh traffic's source `Mesh` and source _tags_.
+
+First of all, we require users to explicitly enable cross-mesh traffic on
+the `MeshGateway` listener as well as mTLS on any meshes doing cross-mesh
+traffic.
+This option will allow us to configure the corresponding Envoy listener
+with all mesh CAs.
```

Comment:
Will we be able to control if the listener is exposed publicly or not?
If the listener is purely private (only for Xmesh) it would be nice to rely on Envoy's load-balancing and not require any extra infra no?

Reply:
Yeah and I actually think a field like `internal: true` that also configures mTLS would be a better option here.

Reply:
@lahabana  When you say no extra infra, you mean rely on Envoy to route traffic directly to the `Dataplane`s serving the `MeshGateway`, correct?

Reply:
Yes correct

Reply:
Any opinions on:

```
listeners:
- port: 8080
  protocol: HTTP
  crossMesh: true
  hostname: gateway.mesh
  tags:
    port: "8080"
```

vs

```
listeners:
- port: 8080
  protocol: HTTPS
  tls:
    mode: CROSS_MESH
  hostname: gateway.mesh
  tags:
    port: "8080"
```

? I prefer the first because if the listener is cross-mesh only, then from the point of view of other services, the protocol is HTTP, it's just protected via mTLS like every innermesh connection. i.e. the way to send a request is with `curl http://gateway.mesh`, not `curl https://gateway.mesh`. I'll add this to the proposal.

Reply:
I 100% agree with you indeed

Reply:
One question are crossMesh and tls mutually exclusive?

Reply:
At the moment there's nothing to configure with `.tls` except the mode. So the only thing to configure would be `protocol: HTTPS` over TLS, I suppose this can make sense but we have to decide whether to support it or just only allow `protocol: HTTP`

Reply:
ok up to you :)